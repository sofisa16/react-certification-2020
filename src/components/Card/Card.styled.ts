import styled from 'styled-components';
import CardMedia from '@material-ui/core/CardMedia';
import Card from '@material-ui/core/Card';

const CardRoot = styled(Card)`
  width: 345px;
  height: 345px;
  margin: 10px;
`;

const CardMediaRoot = styled(CardMedia)`
  height: 140px;
`;

const Styled = {
  CardRoot,
  CardMediaRoot,
};

export default Styled;