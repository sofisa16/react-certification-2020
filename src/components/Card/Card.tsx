import React from 'react';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';
import styled from "styled-components";

const CardRoot = styled(Card)`
  width: 345px;
  margin: 10px;
`;

const CardMediaRoot = styled(CardMedia)`
  height: 140px;
`;

interface MediaCardProps {
  title: string;
  thumbnails: string;
  description: string;
}

function MediaCard(props: MediaCardProps): JSX.Element {
  const { title, thumbnails, description } = props;

  return (
    <CardRoot>
      <CardActionArea>
        <CardMediaRoot
          image={thumbnails}
          title={title}
        />
        <CardContent>
          <Typography gutterBottom variant="h5" component="h2">
            {title}
          </Typography>
          <Typography variant="body2" color="textSecondary" component="p">
            {description}
          </Typography>
        </CardContent>
      </CardActionArea>
    </CardRoot>
  );
}

export default MediaCard;