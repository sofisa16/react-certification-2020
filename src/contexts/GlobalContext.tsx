import React, {useState, useEffect, useReducer} from 'react';
import {ThemeProvider} from 'styled-components';
import NoSsr from '@material-ui/core/NoSsr';
import {
  createMuiTheme,
  ThemeProvider as MuiThemeProvider,
  Theme,
} from '@material-ui/core/styles';
import {initialState, reducer, FavoriteVideos, Action} from './../components/FavoritesView/addToFavorites';
import {AUTH_STORAGE_KEY, AUTH_AVATAR, FAVORITES_VIDEOS} from './../utils/constants';
import {storage} from './../utils/storage';
import { useAuth0 } from "@auth0/auth0-react";

interface GlobalContextValues {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
  darkState: boolean;
  setDarkState: React.Dispatch<React.SetStateAction<boolean>>;
  theme: Theme;
  favoriteVideos: FavoriteVideos;
  dispatchFav: React.Dispatch<Action>;
  authenticated: boolean;
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
  avatar: string;
  setAvatar: React.Dispatch<React.SetStateAction<string>>;
}

interface GlobalContextProviderProps {
  children: React.ReactElement | React.ReactElement[];
}

const GlobalContext = React.createContext<GlobalContextValues>({
  search: '',
  setSearch: () => { return; },
  darkState: false,
  setDarkState: () => { return; },
  theme: createMuiTheme({
    palette: {
      type: 'light',
    }
  }),
  favoriteVideos: {},
  dispatchFav: () => { return; },
  authenticated: false,
  setAuthenticated: () => { return; },
  avatar: '',
  setAvatar: () => { return; },
});

const GlobalContextProvider: React.FunctionComponent<GlobalContextProviderProps> =
  (props: GlobalContextProviderProps) => {
    const [search, setSearch] = useState<string>('wizeline');
    const [darkState, setDarkState] = useState<boolean>(false);
    const [theme, setTheme] = useState<Theme>(
      createMuiTheme({
        palette: {
          type: 'light',
        },
      })
    );
    const [favoriteVideos, dispatchFav] = useReducer(reducer, initialState);
    const [authenticated, setAuthenticated] = useState<boolean>(false);
    const [avatar, setAvatar] = useState<string>('');
    const { user, isAuthenticated } = useAuth0();

    useEffect(
      () => {
        setAuthenticated(isAuthenticated);
      },
      [isAuthenticated]
    );

    useEffect(
      () => {
        if (user) {
          setAvatar(user.picture);
        }
      },
      [user]
    );

    useEffect(
      () => {
        setAuthenticated(Boolean(storage.get(AUTH_STORAGE_KEY)));
        setAvatar(String(storage.get(AUTH_AVATAR)));
        dispatchFav({
          type: 'addAll',
          all: JSON.parse(String(storage.get(FAVORITES_VIDEOS))),
        });
      },
      []
    );

    useEffect(
      () => {
        const palletType = darkState ? 'dark' : 'light';
        setTheme(
          createMuiTheme({
            palette: {
              type: palletType,
            }
          })
        );
      },
      [darkState]
    );

    useEffect(
      () => {
        storage.set(FAVORITES_VIDEOS, JSON.stringify(favoriteVideos));
      },
      [favoriteVideos]
    );

    const value: GlobalContextValues = {
      search,
      setSearch,
      darkState,
      setDarkState,
      theme,
      favoriteVideos,
      dispatchFav,
      authenticated,
      setAuthenticated,
      avatar,
      setAvatar,
    };

    return (
      <NoSsr>
        <MuiThemeProvider theme={theme}>
          <ThemeProvider theme={theme}>
            <GlobalContext.Provider
              value={value}
            >
              {props.children}
            </GlobalContext.Provider>
          </ThemeProvider>
        </MuiThemeProvider>
      </NoSsr>
    );
  };

export {GlobalContextProvider, GlobalContext};