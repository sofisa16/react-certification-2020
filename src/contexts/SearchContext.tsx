import React, {useState} from 'react';

interface SearchContextValues {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
}

interface SearchContextProviderProps {
  children: React.ReactElement | React.ReactElement[];
}

const SearchContext = React.createContext<SearchContextValues>({
  search: "",
  setSearch: () => { return; },
});

const SearchContextProvider: React.FunctionComponent<SearchContextProviderProps> =
  (props: SearchContextProviderProps) => {
    const [search, setSearch] = useState<string>('wizeline');
    const value: SearchContextValues = {
      search,
      setSearch,
    };

    return (
      <SearchContext.Provider
        value={value}
      >
        {props.children}
      </SearchContext.Provider>
    );
  };

export {SearchContextProvider, SearchContext};