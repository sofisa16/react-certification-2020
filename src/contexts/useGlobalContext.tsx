import {useContext} from 'react';
import {GlobalContext, GlobalContextValues} from './GlobalContext';

function useGlobalContext(): GlobalContextValues {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('Please use the "useContext" hook with a GlobalContextProvider.')
  }

  return context;
}

export default useGlobalContext;