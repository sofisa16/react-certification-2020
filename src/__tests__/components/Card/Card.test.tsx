import React from 'react';
import { render } from '@testing-library/react';
import Card from '../../../components/Card/Card';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import {GlobalContextProvider} from './../../../contexts/GlobalContext';
import CssBaseline from '@material-ui/core/CssBaseline';

test('renders Card title', () => {
  const history = createMemoryHistory();
  const route = '/';
  history.push(route);

  const card = render(
    <Router history={history}>
      <Card
        title="title"
        thumbnails="thumbnails"
        description="description"
        key="key"
        videoId="a"
      />
    </Router>
  );
  const title = card.getByText('title');
  expect(title).toBeInTheDocument();
});

test('renders Card description', () => {
  const history = createMemoryHistory();
  const route = '/';
  history.push(route);

  const card = render(
    <Router history={history}>
      <Card
        title="title"
        thumbnails="thumbnails"
        description="description"
        key="key"
        videoId="a"
      />
    </Router>
  );
  const description = card.getByText('description');
  expect(description).toBeInTheDocument();
});

it('renders Card component', () => {
  const history = createMemoryHistory();
  const route = '/';
  history.push(route);

  const tree = render(
    <GlobalContextProvider>
      <CssBaseline>
        <Router history={history}>
          <Card
            title="title"
            thumbnails="thumbnails"
            description="description"
            key="key"
            videoId="a"
          />
        </Router>
      </CssBaseline>
    </GlobalContextProvider>
    );
  expect(tree).toMatchSnapshot();
});