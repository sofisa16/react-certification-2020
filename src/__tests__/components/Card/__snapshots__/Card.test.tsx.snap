// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Card component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="sc-bdfBwQ lclvs"
        href="/videos/a"
      >
        <div
          class="MuiPaper-root MuiCard-root sc-gsTCUz dqPDNH MuiPaper-elevation1 MuiPaper-rounded"
        >
          <button
            class="MuiButtonBase-root MuiCardActionArea-root"
            tabindex="0"
            type="button"
          >
            <div
              alt-text="title"
              class="MuiCardMedia-root sc-dlfnbm dXdMTU"
              style="background-image: url(thumbnails);"
              title="title"
            />
            <div
              class="MuiCardContent-root"
            >
              <h2
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
              >
                title
              </h2>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                description
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="sc-bdfBwQ lclvs"
      href="/videos/a"
    >
      <div
        class="MuiPaper-root MuiCard-root sc-gsTCUz dqPDNH MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            alt-text="title"
            class="MuiCardMedia-root sc-dlfnbm dXdMTU"
            style="background-image: url(thumbnails);"
            title="title"
          />
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
            >
              title
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              description
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
